### PR TITLE
Fix 8407: Allow (weak)map & (weak)set constructor when targeting ES5 using ES6 library

### LINKED ISSUE #8407: Map constructor doesn't work with target ES5, lib ES6
**TypeScript Version:** 

1.7.5 / 1.8.0-beta / nightly (1.9.0-dev.20160430)

**Code**

```ts
const x = new Map([["a", 1]]); // Error
const y = new Map<string, number>([["a", 1]]); // Error
const z = new Map([<[string, number]>["a", 1]]); // Works
```
```json
{
	"compilerOptions": {
		"target": "es5",
		"lib": ["es2016"]
	}
}
```
**Expected behavior:**
No errors

**Actual behavior:** 
Errors are reported on the first two lines:
```
Argument of type '(string | number)[][]' is not assignable to parameter of type 'Iterable<[{}, {}]>'. 
Argument of type '(string | number)[][]' is not assignable to parameter of type 'Iterable<[string, number]>'.
```
I think the reason is that the compiler doesn't use iterators to resolve the type arguments when the target is ES5. Adding an overload to the `Map` constructor, which takes an array of tuples instead of an iterator, would probably fix this. I'm not sure whether this happens on more locations though, if so it might be worth to pick up iterators in the compiler even when the target is ES5.

### PR DESCRIPTION
**Pull request by yuit**
Fix #8407  and move iterable and symbol into correct file

### CODE REVIEW

File: src/lib/es2015.symbol.wellknown.d.ts [hunk 1]
```diff
@@ -124,6 +124,15 @@ interface WeakMap<K, V>{
     readonly [Symbol.toStringTag]: "WeakMap";
 }
 
+interface Set<T> {
+    [Symbol.iterator]():IterableIterator<T>;
```
Comment (mhegazy) [inline src/lib/es2015.symbol.wellknown.d.ts]: should not this be in iterable.d.ts?

File: src/lib/es2015.iterable.d.ts [hunk 2]
```diff
@@ -64,6 +64,22 @@ interface WeakMapConstructor {
...
+    entries(): IterableIterator<[T, T]>;
+    keys(): IterableIterator<T>;
+    values(): IterableIterator<T>;
+}
+
+interface SetConstructor {
+    new <T>(iterable: Iterable<T>): Set<T>;
+}
+
+interface WeakSet<T> { }
```
Comment (mhegazy) [inline src/lib/es2015.iterable.d.ts]: not needed.

**Comment by mhegazy** [comment]
one comment about empty interface declaration, other than that, looks good.

File: src/lib/es2015.iterable.d.ts (hunk 2 above)
Comment (yuit) [inline src/lib/es2015.iterable.d.ts]: It is here because `WeakSetConstructor` return `WeakSet`

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 1 above)
Comment (yuit) [inline src/lib/es2015.symbol.wellknown.d.ts]: I put it here as it uses Symbol.iterator. Though same argument can go with the return type

File: src/lib/es2015.iterable.d.ts (hunk 2 above)
Comment (mhegazy) [inline src/lib/es2015.iterable.d.ts]: but you need the collections for this to work, right?

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 1 above)
Comment (mhegazy) [inline src/lib/es2015.symbol.wellknown.d.ts]: I think iterator is different, and should be in iterables and not in the wellknwon symbols.

File: src/lib/es2015.iterable.d.ts (hunk 2 above)
Comment (yuit) [inline src/lib/es2015.iterable.d.ts]: That is true :sweat_smile:

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 1 above)
Comment (yuit) [inline src/lib/es2015.symbol.wellknown.d.ts]: Just curious why they are different?

Comment (mhegazy) [inline src/lib/es2015.symbol.wellknown.d.ts]: Iterable definition depends on Symbol.Iterator. it is rather meaningless to have iterables and not the definition of Symbol.Iterator.

the other ones are self contained, you can chose to use them or not.

File: src/lib/es2015.iterable.d.ts (hunk 2 above)
Comment (yuit) [inline src/lib/es2015.iterable.d.ts]: talk offline @mhegazy we will keep the empty interface to prevent such behaviour

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**